okhttp/src/main/java/okhttp3/internal/http/RealInterceptorChain.kt

```diff
@@ -33,82 +34,76 @@ import okhttp3.internal.connection.RealCall
+      readTimeoutMillis, writeTimeoutMillis)
 
-  override fun connectTimeoutMillis(): Int = connectTimeout
+  override fun connection(): Connection? = exchange?.connection()
 
-  override fun withConnectTimeout(timeout: Int, unit: TimeUnit): Interceptor.Chain {
-    val millis = checkDuration("timeout", timeout.toLong(), unit)
-    return RealInterceptorChain(interceptors, call, exchange, index, request, millis,
-        readTimeout, writeTimeout)
-  }
+  override fun connectTimeoutMillis(): Int = connectTimeoutMillis
 
-  override fun readTimeoutMillis(): Int = readTimeout
+  override fun withConnectTimeout(timeout: Int, unit: TimeUnit): Interceptor.Chain =
+      copy(connectTimeoutMillis = checkDuration("connectTimeout", timeout.toLong(), unit))
 
-  override fun withReadTimeout(timeout: Int, unit: TimeUnit): Interceptor.Chain {
-    val millis = checkDuration("timeout", timeout.toLong(), unit)
-    return RealInterceptorChain(interceptors, call, exchange, index, request, connectTimeout,
-        millis, writeTimeout)
-  }
+  override fun readTimeoutMillis(): Int = readTimeoutMillis
 
-  override fun writeTimeoutMillis(): Int = writeTimeout
+  override fun withReadTimeout(timeout: Int, unit: TimeUnit): Interceptor.Chain =
+      copy(readTimeoutMillis = checkDuration("readTimeout", timeout.toLong(), unit))
 
-  override fun withWriteTimeout(timeout: Int, unit: TimeUnit): Interceptor.Chain {
-    val millis = checkDuration("timeout", timeout.toLong(), unit)
-    return RealInterceptorChain(interceptors, call, exchange, index, request, connectTimeout,
-        readTimeout, millis)
-  }
+  override fun writeTimeoutMillis(): Int = writeTimeoutMillis
 
-  fun exchange(): Exchange = exchange!!
+  override fun withWriteTimeout(timeout: Int, unit: TimeUnit): Interceptor.Chain =
+      copy(writeTimeoutMillis = checkDuration("writeTimeout", timeout.toLong(), unit))
 
-  override fun call(): RealCall = call
+  override fun call(): Call = call
 
   override fun request(): Request = request
 
-  override fun proceed(request: Request): Response {
-    return proceed(request, exchange)
-  }
-
   @Throws(IOException::class)
-  fun proceed(request: Request, exchange: Exchange?): Response {
-    if (index >= interceptors.size) throw AssertionError()
+  override fun proceed(request: Request): Response {
+    check(index < interceptors.size)
 
     calls++
 
-    // If we already have an exchange, confirm that the incoming request will use it.
-    check(this.exchange == null || this.exchange.connection()!!.supportsUrl(request.url)) {
-      "network interceptor ${interceptors[index - 1]} must retain the same host and port"
-    }
-
-    // If we already have an exchange, confirm that this is the only call to chain.proceed().
-    check(this.exchange == null || calls <= 1) {
-      "network interceptor ${interceptors[index - 1]} must call proceed() exactly once"
+    if (exchange != null) {
```

Comment:
For documentation purposes might be worth extracting this condition into an "isNetworkInterceptor" condition

Reply:
I like it. But I tried it, and then it makes dereferencing `exchange` more difficult because Kotlin can’t infer it’s non-null.

Reply:
We can fix when contracts stabilise